src/components/MagicCodeInput.js

```diff
@@ -0,0 +1,347 @@
+            return;
+        }
+        clearTimeout(this.focusTimeout);
+    }
+
+    /**
+     * Focuses on the input when it is pressed.
+     *
+     * @param {Object} event
+     * @param {Number} index
+     */
+    onFocus(event) {
+        event.preventDefault();
+        this.setState({
+            input: '',
+        });
+    }
+
+    /**
+     * Callback for the onPress event, updates the indexes
+     * of the currently focused input.
+     *
+     * @param {Object} event
+     * @param {Number} index
+     */
+    onPress(event, index) {
+        event.preventDefault();
+        this.setState({
+            input: '',
+            focusedIndex: index,
+            editIndex: index,
+        });
+    }
+
+    /**
+     * Updates the magic inputs with the contents written in the
+     * input. It spreads each number into each input and updates
+     * the focused input on the next empty one, if exists.
+     * It handles both fast typing and only one digit at a time
+     * in a specific position.
+     *
+     * @param {String} value
+     */
+    onChangeText(value) {
+        if (_.isUndefined(value) || _.isEmpty(value) || !ValidationUtils.isNumeric(value)) {
+            return;
+        }
+
+        this.setState((prevState) => {
+            const numbersArr = value.trim().split('').slice(0, CONST.MAGIC_CODE_LENGTH - prevState.editIndex);
+            const numbers = [
+                ...prevState.numbers.slice(0, prevState.editIndex),
+                ...numbersArr,
+                ...prevState.numbers.slice(numbersArr.length + prevState.editIndex, CONST.MAGIC_CODE_LENGTH),
+            ];
+
+            // Updates the focused input taking into consideration the last input
+            // edited and the number of digits added by the user.
+            const focusedIndex = Math.min(prevState.editIndex + (numbersArr.length - 1) + 1, CONST.MAGIC_CODE_LENGTH - 1);
+
+            return {
+                numbers,
+                focusedIndex,
+                input: value,
+            };
+        }, () => {
+            const finalInput = this.composeToString(this.state.numbers);
+            this.props.onChangeText(finalInput);
+
+            // Blurs the input and removes focus from the last input and, if it should submit
+            // on complete, it will call the onFulfill callback.
+            if (this.props.shouldSubmitOnComplete && _.filter(this.state.numbers, n => ValidationUtils.isNumeric(n)).length === CONST.MAGIC_CODE_LENGTH) {
+                this.inputRefs[this.state.editIndex].blur();
+                this.setState({focusedIndex: undefined}, () => this.props.onFulfill(finalInput));
+            }
+        });
+    }
+
+    /**
+     * Handles logic related to certain key presses.
+     *
+     * NOTE: when using Android Emulator, this can only be tested using
+     * hardware keyboard inputs.
+     *
+     * @param {Object} event
+     */
+    onKeyPress({nativeEvent: {key: keyValue}}) {
```

Comment:
Methods should never be named for what they handle, they should only be named for what they do. This is part of the checklist:

> I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. toggleReport and not onIconClick)

Reply:
My bad, I must have skipped it when I read it. There's an issue regarding MagicCodeInput yet - https://github.com/Expensify/App/issues/18325 - where this can be fixed, if I end up assigned, I can do it there.